perf(oxlint/lsp): store only `LinterCodeAction` instead of the complete `DiagnosticReport`

Sysix opened the pull request:
> This PR implements a performance optimization for the LSP linter by storing only LinterCodeAction objects in the cache instead of complete DiagnosticReport objects. This reduces memory usage since diagnostics are only needed when initially reported, while code actions are needed repeatedly for quick fixes and "fix all" operations.

Sysix commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.com/github/pr/oxc-project/oxc/16516?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#16516** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/16516?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/oxc-project/oxc/16516?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#16515** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/16515?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#16514** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/16514?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#16513** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/16513?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=16516).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

camc314 commented:
### Merge activity

* **Dec 5, 1:47 PM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=oxc) once it meets the requirements.
* **Dec 5, 1:47 PM UTC**: `camc314` added this pull request to the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=oxc).

Sysix committed:
perf(oxlint/lsp): store only `LinterCodeAction` instead of the complete `DiagnosticReport` (#16516)

> This PR implements a performance optimization for the LSP linter by storing only LinterCodeAction objects in the cache instead of complete DiagnosticReport objects. This reduces memory usage since diagnostics are only needed when initially reported, while code actions are needed repeatedly for quick fixes and "fix all" operations.
